Patch by MSP-Greg
1st hunk adjusts timeout for long disk write
2nd hunk may lower false failures from test_write_no_garbage when run parallel
diff --git a/testdiff --git a/test/ruby/test_io.rb b/test/ruby/test_io.rb
index 1c45787648..12f14e81ef 100644
--- a/test/ruby/test_io.rb
+++ b/test/ruby/test_io.rb
@@ -3403,9 +3403,9 @@ def t.close(unlink_now = false)
         end
         ret
       end
-
+      time_out = /mingw/ =~ RUBY_PLATFORM ? 50 : 30  # Appveyor 2017-11
       begin
-        assert_separately(["-", bug8431, t.path], <<-"end;", timeout: 30)
+        assert_separately(["-", bug8431, t.path], <<-"end;", timeout: time_out)
           msg = ARGV.shift
           f = open(ARGV[0], "wb")
           f.seek(0xffff_ffff)
@@ -3705,25 +3705,43 @@ def test_closed_stream_in_rescue
     def test_write_no_garbage
       skip "multiple threads already active" if Thread.list.size > 1
       res = {}
+      before = []
+      after = []
+      freeze = []
+      a_n = []
+      a_s = []
+      a_buf = []
       ObjectSpace.count_objects(res) # creates strings on first call
       [ 'foo'.b, '*' * 24 ].each do |buf|
         with_pipe do |r, w|
           GC.disable
           begin
-            before = ObjectSpace.count_objects(res)[:T_STRING]
-            n = w.write(buf)
-            s = w.syswrite(buf)
-            after = ObjectSpace.count_objects(res)[:T_STRING]
+            before << ObjectSpace.count_objects(res)[:T_STRING]
+            a_n    << w.write(buf)
+            a_s    << w.syswrite(buf)
+            after  << ObjectSpace.count_objects(res)[:T_STRING]
           ensure
             GC.enable
           end
-          assert_equal before, after,
-            "no strings left over after write [ruby-core:78898] [Bug #13085]: #{ before } strings before write -> #{ after } strings after write"
-          assert_not_predicate buf, :frozen?, 'no inadvertent freeze'
-          assert_equal buf.bytesize, n, 'IO#write wrote expected size'
-          assert_equal s, n, 'IO#syswrite wrote expected size'
+          a_buf  << buf.bytesize          
+          freeze << buf.frozen?
         end
       end
+      0.upto(before.length() -1) { |i|
+        assert(before[i] >= after[i],
+          "no strings left over after write [ruby-core:78898] [Bug #13085]: #{ before[i] } strings before write -> #{ after[i] } strings after write")
+        assert !freeze[i], 'no inadvertent freeze'
+        assert_equal a_buf[i], a_n[i], 'IO#write wrote expected size'
+        assert_equal a_s[i]  , a_n[i], 'IO#syswrite wrote expected size'
+      }
+      # clear for good measure
+      res = {}
+      before = []
+      after = []
+      freeze = []
+      a_n = []
+      a_s = []
+      a_buf = []
     end
 
     def test_pread
